File: 10-greedy/01_buy_sell_stock_ii.py
# ============================================================
# PATTERN: GREEDY
# ============================================================
# Greedy makes the locally optimal choice at each step, hoping for a global optimum. Works when:
# 1. **Greedy choice property** -- local optimum leads to global optimum
# 2. **Optimal substructure** -- optimal solution contains optimal sub-solutions
#
# **Common strategies:**
# - Sort by some criteria (end time, ratio, etc.) then scan
# - Two pointers from both ends
# - Track running max/min/sum
#
# **How to verify greedy works:** Try to construct a counterexample. If you can't, it's likely correct.
#
# ---
# ============================================================

import sys, os; sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))
from test_runner import run_tests


# Best Time to Buy and Sell Stock II (LC #122) -- Medium
# Max profit with unlimited transactions.
#
#   Key: Collect every upward slope: if `prices[i] > prices[i-1]`, add the difference

def max_profit(prices):
    pass


run_tests(max_profit, [
    (([7,1,5,3,6,4],), 7),
    (([1,2,3,4,5],),   4),
    (([7,6,4,3,1],),   0),
])

File: 10-greedy/01_buy_sell_stock_ii.c
#include "../test_runner.h"

int max_profit(int* prices, int n) {
    return -1;
}

int main() {
    TEST_INT("ups and downs", max_profit(ARR(7,1,5,3,6,4), 6), 7);
    TEST_INT("all up",        max_profit(ARR(1,2,3,4,5), 5),   4);
    TEST_INT("all down",      max_profit(ARR(7,6,4,3,1), 5),   0);
    TEST_SUMMARY();
}
